Supported collection types doc for System.Text.Json (#22217)
dotnet/docs: docs/standard/serialization/system-text-json-supported-collection-types.md

---
title: "Supported collection types in System.Text.Json"
description: "Learn which collection types are supported for serialization by the APIs in the System.Text.Json namespace."
ms.date: 01/04/2021

> no-loc: [System.Text.Json]

Reviewer: ```suggestion
no-loc: [System.Text.Json]
ms.topic: reference
```

zone_pivot_groups: dotnet-version
helpviewer_keywords:
  - "JSON serialization"
  - "serializing objects"
  - "serialization"
  - "objects, serializing"
---

# Supported collection types in System.Text.Json

This article gives an overview of which collections are supported for serialization and deserialization. <xref:System.Text.Json.JsonSerializer?displayProperty=nameWithType> supports a collection type for serialization if it:

* Derives from <xref:System.Collections.IEnumerable>.
* Contains elements that are serializable.

The serializer calls the <xref:System.Collections.IEnumerable.GetEnumerator> method, and writes the elements.

Deserialization is more complicated and is not supported for some collection types.

> ## System.Collections namespace

Reviewer: Maybe give an intro here to what the subsections are? The following sections are organized by namespace and show which types are supported for serialization and deserialization...

| [KeyedCollection\<string, TValue>](xref:System.Collections.ObjectModel.KeyedCollection%602) |✔️|❌|
| <xref:System.Collections.ObjectModel.ObservableCollection%601> | ✔️            | ✔️             |
| <xref:System.Collections.ObjectModel.ReadOnlyCollection%601>   | ✔️            | ❌             |
| <xref:System.Collections.ObjectModel.ReadOnlyDictionary%602>   | ✔️            | ❌             |
| <xref:System.Collections.ObjectModel.ReadOnlyObservableCollection%601> | ✔️    | ❌             |

::: zone-end

::: zone pivot="dotnet-core-3-1"

| Type                                                           | Serialization | Deserialization |
|----------------------------------------------------------------|---------------|-----------------|
| <xref:System.Collections.ObjectModel.Collection%601>           | ✔️            | ✔️             |
| [KeyedCollection\<string, TValue>](xref:System.Collections.ObjectModel.KeyedCollection%602) |✔️|❌|
| <xref:System.Collections.ObjectModel.ObservableCollection%601> | ✔️            | ✔️             |
| <xref:System.Collections.ObjectModel.ReadOnlyCollection%601>   | ✔️            | ❌             |

> | [ReadOnlyDictionary\<string, TValue>](xref:System.Collections.ObjectModel.ReadOnlyDictionary%602) |✔️|❌|

Reviewer: Maybe call out somewhere that pre-.NET 5, the key type has to be string?

| <xref:System.Collections.ObjectModel.ReadOnlyObservableCollection%601>| ✔️     | ❌             |

::: zone-end

## Custom collections

> Any collection type that isn't in one of the preceding namespaces is considered a custom collection. Such types include user-defined types and types defined by ASP.NET Core. For example, <xref:Microsoft.Extensions.Primitives?displayProperty=fullName> is in this group.

Reviewer: Would a dictionary whose key type isn't string fall under this category?

Author: No, because it's in the preceding namespaces, it's just not supported in 3.1.

  * <xref:System.Collections.IDictionary>
  * <xref:System.Collections.Generic.IDictionary%602>
  * <xref:System.Collections.IList>
  * <xref:System.Collections.Generic.IList%601>
  * <xref:System.Collections.Queue>
  * <xref:System.Collections.Generic.Queue%601>
  * <xref:System.Collections.Stack> \*
  * <xref:System.Collections.Generic.Stack%601> \*
* Contains element types that are supported by <xref:System.Text.Json.JsonSerializer>.

::: zone-end

::: zone pivot="dotnet-core-3-1"

* Isn't an interface or abstract.
* Has a parameterless constructor.
* Implements or inherits one or more of the following interfaces or classes:
  * <xref:System.Collections.Concurrent.ConcurrentQueue%601>
  * <xref:System.Collections.Concurrent.ConcurrentStack%601> \*
  * <xref:System.Collections.Generic.ICollection%601>
  * <xref:System.Collections.IDictionary>
  * [IDictionary\<string, TValue>](xref:System.Collections.Generic.IDictionary%602)
  * <xref:System.Collections.IList>
  * <xref:System.Collections.Generic.IList%601>
  * <xref:System.Collections.Queue>
  * <xref:System.Collections.Generic.Queue%601>
  * <xref:System.Collections.Stack> \*
  * <xref:System.Collections.Generic.Stack%601> \*

> * Contains element types that are supported by <xref:System.Text.Json.JsonSerializer>.

Reviewer: Consider putting this bullet before "Implements or inherits..." for readability and so the asterisk matches up better.

Author: Good idea, thanks!

::: zone-end

\* See [Support round trip for Stack\<T>](system-text-json-converters-how-to.md#support-round-trip-for-stackt).

### Custom collections with known issues

> There are known issues with some custom collections:

Reviewer: ```suggestion
There are known issues with the following custom collections:
```

- <xref:System.Dynamic.ExpandoObject>: See [dotnet/runtime#29690](https://github.com/dotnet/runtime/issues/29690).
- <xref:System.Dynamic.DynamicObject>: See [dotnet/runtime#1808](https://github.com/dotnet/runtime/issues/1808).
- <xref:System.Data.DataTable>: See [dotnet/docs#21366](https://github.com/dotnet/docs/issues/21366).
- <xref:Microsoft.AspNetCore.Http.FormFile?displayProperty=fullName>: See [dotnet/runtime#1559](https://github.com/dotnet/runtime/issues/1559).
- <xref:Microsoft.AspNetCore.Http.IFormCollection?displayProperty=fullName>: See [dotnet/runtime#1559](https://github.com/dotnet/runtime/issues/1559).

## See also

> For more information, see the [open issues in System.Text.Json](https://github.com/dotnet/runtime/issues?q=is%3Aopen+is%3Aissue+label%3Aarea-System.Text.Json).

Reviewer: For more information about ...

  * <xref:System.Collections.Generic.IDictionary%602>
  * <xref:System.Collections.IList>
  * <xref:System.Collections.Generic.IList%601>
  * <xref:System.Collections.Queue>
  * <xref:System.Collections.Generic.Queue%601>
  * <xref:System.Collections.Stack> \*
  * <xref:System.Collections.Generic.Stack%601> \*

::: zone-end

::: zone pivot="dotnet-core-3-1"

* Isn't an interface or abstract.
* Has a parameterless constructor.
* Contains element types that are supported by <xref:System.Text.Json.JsonSerializer>.
* Implements or inherits one or more of the following interfaces or classes:
  * <xref:System.Collections.Concurrent.ConcurrentQueue%601>
  * <xref:System.Collections.Concurrent.ConcurrentStack%601> \*
  * <xref:System.Collections.Generic.ICollection%601>
  * <xref:System.Collections.IDictionary>
  * [IDictionary\<string, TValue>](xref:System.Collections.Generic.IDictionary%602) \*\*
  * <xref:System.Collections.IList>
  * <xref:System.Collections.Generic.IList%601>
  * <xref:System.Collections.Queue>
  * <xref:System.Collections.Generic.Queue%601>
  * <xref:System.Collections.Stack> \*
  * <xref:System.Collections.Generic.Stack%601> \*

> \*\* Non-`string` keys are not supported in .NET Core 3.1.

Reviewer: ```suggestion
  \*\* Non-`string` keys are not supported in .NET Core 3.1.
```

::: zone-end

> \* See [Support round trip for Stack\<T>](system-text-json-converters-how-to.md#support-round-trip-for-stackt).

Reviewer: ```suggestion
  \* See [Support round trip for Stack\<T>](system-text-json-converters-how-to.md#support-round-trip-for-stackt).
```

>   \* See [Support round trip for Stack\<T>](system-text-json-converters-how-to.md#support-round-trip-for-stackt).

Reviewer: It didn't indent properly, I guess because it's separated from the zone. You might add is separately to each zone so it 1) is indented and 2) appears before the double asterisk.

### Custom collections with known issues

There are known issues with the following custom collections:

- <xref:System.Dynamic.ExpandoObject>: See [dotnet/runtime#29690](https://github.com/dotnet/runtime/issues/29690).
- <xref:System.Dynamic.DynamicObject>: See [dotnet/runtime#1808](https://github.com/dotnet/runtime/issues/1808).
- <xref:System.Data.DataTable>: See [dotnet/docs#21366](https://github.com/dotnet/docs/issues/21366).
- <xref:Microsoft.AspNetCore.Http.FormFile?displayProperty=fullName>: See [dotnet/runtime#1559](https://github.com/dotnet/runtime/issues/1559).
- <xref:Microsoft.AspNetCore.Http.IFormCollection?displayProperty=fullName>: See [dotnet/runtime#1559](https://github.com/dotnet/runtime/issues/1559).

## See also

> For more information about known issues, see the [open issues in System.Text.Json](https://github.com/dotnet/runtime/issues?q=is%3Aopen+is%3Aissue+label%3Aarea-System.Text.Json).

Reviewer: Shouldn't this sentence be moved to the end of the previous section about known issues?

> ## See also
> 
> For more information about known issues, see the [open issues in System.Text.Json](https://github.com/dotnet/runtime/issues?q=is%3Aopen+is%3Aissue+label%3Aarea-System.Text.Json).

Author: ```suggestion
For more information about known issues, see the [open issues in System.Text.Json](https://github.com/dotnet/runtime/issues?q=is%3Aopen+is%3Aissue+label%3Aarea-System.Text.Json).

## See also
```
